generate/genfilehcl/genfile_test.go

```diff
@@ -0,0 +1,1327 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genfilehcl_test
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/madlambda/spells/assert"
+	"github.com/mineiros-io/terramate/config"
+	"github.com/mineiros-io/terramate/generate/genfilehcl"
+	"github.com/mineiros-io/terramate/hcl/eval"
+	"github.com/mineiros-io/terramate/test"
+	"github.com/mineiros-io/terramate/test/hclwrite"
+	"github.com/mineiros-io/terramate/test/sandbox"
+	"github.com/rs/zerolog"
+)
+
+func TestLoadGeneratedHCL(t *testing.T) {
+	type (
+		hclconfig struct {
+			path     string
+			filename string
+			add      fmt.Stringer
+		}
+		genFile struct {
+			body   fmt.Stringer
+			origin string
+		}
+		result struct {
+			name string
+			hcl  genFile
+		}
+		testcase struct {
+			name    string
+			stack   string
+			configs []hclconfig
+			want    []result
+			wantErr error
+		}
+	)
+
+	defaultCfg := func(dir string) string {
+		return filepath.Join(dir, config.DefaultFilename)
+	}
+
+	tcases := []testcase{
+		{
+			name:  "generate file with only text in it",
+			stack: "/stack",
+			configs: []hclconfig{
+				{
+					path: "/stack",
+					add: generateHCL(
+						labels("test.txt"),
+						str("content", "something = \"hello world\""),
+					),
+				},
+			},
+			want: []result{
+				{
+					name: "test.txt",
+					hcl: genFile{
+						origin: defaultCfg("/stack"),
+						body: hcldoc(
+							str("something", "hello world"),
+						),
+					},
+				},
+			},
+		},
+		{
+			name:  "generate file with only text in it",
+			stack: "/stack",
+			configs: []hclconfig{
+				{
+					path: "/stack",
+					add: generateHCL(
+						labels("test.txt"),
+						str("content", "something = \"something\""),
+					),
+				},
+			},
+			want: []result{
+				{
+					name: "test.txt",
+					hcl: genFile{
+						origin: defaultCfg("/stack"),
+						body: hcldoc(
+							str("something", "something"),
+						),
+					},
+				},
+			},
+		},
+		{
+			name:  "generate file with function",
+			stack: "/stack",
+			configs: []hclconfig{
+				{
+					path: "/stack",
+					add: generateHCL(
+						labels("test.json"),
+						expr("content", "\"something = tm_max(5, 12, 9)\""),
+					),
+				},
+			},
+			want: []result{
```

Comment:
Kinda odd that the tests "want" an "hcl", the idea of generate_file is to generate random files (JSON/YAML/some text/etc) :thinking: .